File: 0862-shortest-subarray-with-sum-at-least-k/0862-shortest-subarray-with-sum-at-least-k.c
#include <stdlib.h>
#include <limits.h>

/**
 * @brief Finds the length of the shortest subarray with a sum >= k.
 *
 * @param nums (int*): Array of integers.
 * @param numsSize (int): Size of the nums array.
 * @param k (int): Target sum.
 *
 * @return (int): Length of the shortest subarray with a sum >= k, or -1 if no such subarray exists.
 */
int shortestSubarray(int* nums, int numsSize, int k) {
    // Create the prefix sum array
    long* prefixSum = (long*)malloc((numsSize + 1) * sizeof(long));
    if (!prefixSum) return -1;  // Handle memory allocation failure

    prefixSum[0] = 0;
    for (int i = 0; i < numsSize; i++) {
        prefixSum[i + 1] = prefixSum[i] + nums[i];
    }

    // Initialize a deque to store indices
    int* deque = (int*)malloc((numsSize + 1) * sizeof(int));
    if (!deque) {
        free(prefixSum);
        return -1;
    }
    int dequeStart = 0, dequeEnd = 0;

    int result = INT_MAX;  // Initialize result to infinity (no valid subarray found)

    // Iterate over the prefix sum array
    for (int i = 0; i <= numsSize; i++) {
        // Remove elements from the front of the deque if they form a valid subarray
        while (dequeStart < dequeEnd && prefixSum[i] - prefixSum[deque[dequeStart]] >= k) {
            result = (i - deque[dequeStart] < result) ? i - deque[dequeStart] : result;
            dequeStart++;
        }

        // Maintain monotonicity: remove elements from the back if they are greater than or equal to the current prefix sum
        while (dequeStart < dequeEnd && prefixSum[deque[dequeEnd - 1]] >= prefixSum[i]) {
            dequeEnd--;
        }

        // Add the current index to the deque
        deque[dequeEnd++] = i;
    }

    // Free allocated memory
    free(prefixSum);
    free(deque);

    // Return the result, or -1 if no valid subarray was found
    return result == INT_MAX ? -1 : result;
}

File: Difficulty/C. Hard/0862-shortest-subarray-with-sum-at-least-k/0862-shortest-subarray-with-sum-at-least-k.py
from typing import List
from collections import deque

class Solution:
    def shortestSubarray(self, nums: List[int], k: int) -> int:
        """
        Finds the length of the shortest subarray with a sum >= k.

        Parameters:
        nums (List[int]): List of integers.
        k (int): Target sum.

        Returns:
        int: Length of the shortest subarray with a sum >= k, or -1 if no such subarray exists.
        """
        # Initialize variables
        n = len(nums)
        prefix_sum = [0] * (n + 1)  # Prefix sum array
        for i in range(n):
            prefix_sum[i + 1] = prefix_sum[i] + nums[i]

        result = float("inf")  # Initialize result with infinity (indicating no valid subarray found)
        deque_indices = deque()  # Monotonic queue to store indices

        # Iterate over prefix sums
        for i in range(n + 1):
            # Remove elements from the front of the deque if a valid subarray is found
            while deque_indices and prefix_sum[i] - prefix_sum[deque_indices[0]] >= k:
                result = min(result, i - deque_indices.popleft())

            # Maintain the monotonicity of the deque
            while deque_indices and prefix_sum[i] <= prefix_sum[deque_indices[-1]]:
                deque_indices.pop()

            # Add the current index to the deque
            deque_indices.append(i)

        # Return the result if a valid subarray is found, otherwise -1
        return result if result != float("inf") else -1
